### PR TITLE
Add support for AWS CodeCommit git repositories

### PR DESCRIPTION
**Pull request by wleepang**
See issue #61 

* Adds new RepositoryProvider for AWS CodeCommit
* Updates scm/ProviderConfig to include AWS CodeCommit
* Updates scm/AssetManager to handle AWS CodeCommit repositories

### CODE REVIEW

File: modules/nextflow/src/main/groovy/nextflow/Const.groovy [hunk 1]
```diff
@@ -57,12 +57,12 @@ class Const {
...
      * The app build time as linux/unix timestamp
      */
-    static public final long APP_TIMESTAMP = 1587316425933
+    static public final long APP_TIMESTAMP = 1588567650633
 
     /**
      * The app build number
      */
-    static public final int APP_BUILDNUM = 5333
+    static public final int APP_BUILDNUM = 5426
```
[the remark this answers is not in the document]
Comment (wleepang) [reply modules/nextflow/src/main/groovy/nextflow/Const.groovy]: done

File: modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy [hunk 2]
```diff
@@ -54,7 +54,6 @@ import static nextflow.Const.MANIFEST_FILE_NAME
  */
 
 @Slf4j
-@CompileStatic
 class AssetManager {
```
[the remark this answers is not in the document]
Comment (wleepang) [reply modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy]: Done

File: modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy [hunk 3]
```diff
@@ -944,6 +965,9 @@ class AssetManager {
         // call submodule init
         init.call()
         // call submodule update
+        if( provider.name == "CodeCommit" ) 
+            update.setCredentialsProvider( provider.getAwsCodeCommitCredentialProvider() )
+
         if( provider.hasCredentials() )
             update.setCredentialsProvider( new UsernamePasswordCredentialsProvider(provider.user, provider.password) )
```
[the remark this answers is not in the document]
Comment (wleepang) [reply modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy]: Done

File: modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy [hunk 4]
```diff
@@ -87,6 +90,14 @@ abstract class RepositoryProvider {
      */
     abstract String getRepositoryUrl()
 
+    /**
+     * @return a org.eclipse.jgit.transport.CredentialsProvider object for authenticating git operations
+     * like clone, fetch, pull, and update
+     **/
+    protected CredentialsProvider getGitCredentials() {
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy:97]: We can make this public (which is the default permission in groovy) therefore `protected` can be omitted.

**Comment by wleepang** [comment]
@pditommaso - is it possible to run tests locally?

**Comment by pditommaso** [comment]
Yes, use `make test`.

**Comment by wleepang** [comment]
Is it possible to run a specific test, or just tests within `nextflow.scm`?

**Comment by wleepang** [comment]
n/m - I think I figured it out:

```
make test module=nextflow class=*AwsCodeCommitRepositoryProvider*
```

File: modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy (hunk 4 above)
Comment (wleepang) [reply modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy]: done

**Comment by wleepang** [comment]
Not sure about the differences in the Travis-CI test run vs local.  All tests pass locally except the following for `AssetManager`:

```
        when:
        manager = new AssetManager()
        manager.providerConfigs.add( new ProviderConfig('local-scm', [platform: 'github', server: 'http://foo.bar.com']) )
        result = manager.resolveNameFromGitUrl('https://foo.bar.com/project/xyz.git')
        then:
        result == 'project/xyz'
        manager.hub == 'local-scm'
```

**Comment by wleepang** [comment]
figured out the test variance - lack of aws credentials in Travis

**Comment by wleepang** [comment]
@pditommaso - looks like everything is working

**Comment by pditommaso** [comment]
Hi Lee, thanks a lot for looking into this. One question why the http connection with git-remote-codecommit was chosen instead of plain [http protocol](https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-gc.html)? The latter seems to not depend on the installations of external tools.

**Comment by wleepang** [comment]
The integration doesn't require `git-remote-codecommit` to be installed. It's just borrowing the repository URL pattern since it's simpler to remember and type, and provides more flexibility (i.e. enables named profiles for repository access).

**Comment by pditommaso** [comment]
Still thinking http would fit better to current NF schema. Consider that you don't need to explicitly use the full url to use repo other than GitHub. It can be specified with the `-hub` option. The region could be inferred using the usual `AWS_DEFAULT_REGION` or using adding it as an option in the `scm` config file options  for codecommit.

**Comment by wleepang** [comment]
It is using http under the hood. The http clone url depends on the region a repository is stored in and looks like:

```
https://git-codecommit.<region>.amazonaws.com/v1/repos/<reponame>
```

This url is generated based on the GRC formatted project url.

It is worth considering use cases where you want to run a workflow in `region1` but the repo for the workflow is in `region2`. The inferred region from the DefaultProvider chain wouldn't apply. You would need to specify the correctly regionalized clone url.

Also consider running locally and you have configured named profiles for accessing repos using different roles or in different accounts. This cannot be specified with the http clone url alone, but can be with GRC formatted urls.

So the cases are as follows:

* codecommit://project
  * project repository in the default profile's default region
  * project accessible with the default profile's credentials
  * clone `https://git-codecommit.<default-region>.amazonaws.com/v1/repos/project`
  * use default credentials from `nextflow.config` or AWS Default Credentials Provider Chain
* codecommit::specific-region://project
  * project repository in a region different than the current region or the default profile's default region
  * project accessible with the default profile's credentials
  * clone `https://git-codecommit.specific-region.amazonaws.com/v1/repos/project`
  * use default credentials from either `nextflow.config` or AWS Default Credentials Provider Chain 
* codecommit://profile-name@project
  * project repository in the default profile's default region
  * project accessible with different credentials than the default profile
  * clone `https://git-codecommit.<default-region>.amazonaws.com/v1/repos/project`
  * use credentials for `profile-name` from AWS Profile Credentials Provider (e.g. from `~/.aws/config`)
* codecommit::specific-region://profile-name@project
  * project repository in a region different than the current region or the default profile's default region
  * project accessible with different credentials than the default profile
  * clone `https://git-codecommit.specific-region.amazonaws.com/v1/repos/project`
  * use credentials for `profile-name` from AWS Profile Credentials Provider (e.g. from `~/.aws/config`)

Using the GRC format is already a compact form of the http url you would otherwise need to specify for a repo in CodeCommit, and the heavy lifting to convert the GRC clone url to the correct http clone url and handling alternate profile auth is done by the `AwsCodeCommitRepositoryProvider`. Probably a good thing to add would be the ability to use the http clone url directly with the caveat that it might restrict the use of named profiles.

**Comment by pditommaso** [comment]
I understand that GRC format is more flexible, however, the use of the `AwsCodeCommitRepositoryProvider ` adds too many dependencies in the NF core distribution (see below). This will bloat the installer with a large number of downloads, 99% of the times for average users. 

My understanding is that instead, the plain http protocol would not even require the `java-sdk-codecommit` library. This is why it should be preferred. We can consider adding the support for GCR format once we'll have a plugin system for NF that would to extends core functionality via opt-in extensions. 

```
|    +--- org.springframework.cloud:spring-cloud-config-server:2.2.2.RELEASE
|    |    +--- org.springframework.cloud:spring-cloud-config-client:2.2.2.RELEASE
|    |    |    +--- org.springframework.boot:spring-boot-autoconfigure:2.2.5.RELEASE
|    |    |    |    \--- org.springframework.boot:spring-boot:2.2.5.RELEASE
|    |    |    |         +--- org.springframework:spring-core:5.2.4.RELEASE
|    |    |    |         |    \--- org.springframework:spring-jcl:5.2.4.RELEASE
|    |    |    |         \--- org.springframework:spring-context:5.2.4.RELEASE
|    |    |    |              +--- org.springframework:spring-aop:5.2.4.RELEASE
|    |    |    |              |    +--- org.springframework:spring-beans:5.2.4.RELEASE
|    |    |    |              |    |    \--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |    |              |    \--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |    |              +--- org.springframework:spring-beans:5.2.4.RELEASE (*)
|    |    |    |              +--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |    |              \--- org.springframework:spring-expression:5.2.4.RELEASE
|    |    |    |                   \--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |    +--- org.springframework.cloud:spring-cloud-commons:2.2.2.RELEASE
|    |    |    |    \--- org.springframework.security:spring-security-crypto:5.2.2.RELEASE
|    |    |    +--- org.springframework.cloud:spring-cloud-context:2.2.2.RELEASE
|    |    |    |    \--- org.springframework.security:spring-security-crypto:5.2.2.RELEASE
|    |    |    +--- org.springframework:spring-web:5.2.4.RELEASE
|    |    |    |    +--- org.springframework:spring-beans:5.2.4.RELEASE (*)
|    |    |    |    \--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
|    |    |    \--- com.fasterxml.jackson.core:jackson-databind:2.10.2 (*)
|    |    +--- org.springframework.boot:spring-boot-starter-actuator:2.2.5.RELEASE
|    |    |    +--- org.springframework.boot:spring-boot-starter:2.2.5.RELEASE
|    |    |    |    +--- org.springframework.boot:spring-boot:2.2.5.RELEASE (*)
|    |    |    |    +--- org.springframework.boot:spring-boot-autoconfigure:2.2.5.RELEASE (*)
|    |    |    |    +--- org.springframework.boot:spring-boot-starter-logging:2.2.5.RELEASE
|    |    |    |    |    +--- ch.qos.logback:logback-classic:1.2.3 (*)
|    |    |    |    |    +--- org.apache.logging.log4j:log4j-to-slf4j:2.12.1
|    |    |    |    |    |    +--- org.slf4j:slf4j-api:1.7.25 -> 1.7.30
|    |    |    |    |    |    \--- org.apache.logging.log4j:log4j-api:2.12.1
|    |    |    |    |    \--- org.slf4j:jul-to-slf4j:1.7.30 (*)
|    |    |    |    +--- jakarta.annotation:jakarta.annotation-api:1.3.5
|    |    |    |    +--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |    |    \--- org.yaml:snakeyaml:1.25
|    |    |    +--- org.springframework.boot:spring-boot-actuator-autoconfigure:2.2.5.RELEASE
|    |    |    |    +--- org.springframework.boot:spring-boot-actuator:2.2.5.RELEASE
|    |    |    |    |    +--- org.springframework.boot:spring-boot:2.2.5.RELEASE (*)
|    |    |    |    |    \--- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.10.2
|    |    |    |    |         +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
|    |    |    |    |         +--- com.fasterxml.jackson.core:jackson-core:2.10.2
|    |    |    |    |         \--- com.fasterxml.jackson.core:jackson-databind:2.10.2 (*)
|    |    |    |    +--- org.springframework.boot:spring-boot-autoconfigure:2.2.5.RELEASE (*)
|    |    |    |    +--- com.fasterxml.jackson.core:jackson-databind:2.10.2 (*)
|    |    |    |    +--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |    |    +--- org.springframework:spring-context:5.2.4.RELEASE (*)
|    |    |    |    \--- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.10.2 (*)
|    |    |    \--- io.micrometer:micrometer-core:1.3.5
|    |    |         +--- org.hdrhistogram:HdrHistogram:2.1.11
|    |    |         \--- org.latencyutils:LatencyUtils:2.0.3
|    |    +--- org.springframework.boot:spring-boot-starter-web:2.2.5.RELEASE
|    |    |    +--- org.springframework.boot:spring-boot-starter:2.2.5.RELEASE (*)
|    |    |    +--- org.springframework.boot:spring-boot-starter-json:2.2.5.RELEASE
|    |    |    |    +--- org.springframework.boot:spring-boot-starter:2.2.5.RELEASE (*)
|    |    |    |    +--- org.springframework:spring-web:5.2.4.RELEASE (*)
|    |    |    |    +--- com.fasterxml.jackson.core:jackson-databind:2.10.2 (*)
|    |    |    |    +--- com.fasterxml.jackson.datatype:jackson-datatype-jdk8:2.10.2
|    |    |    |    |    +--- com.fasterxml.jackson.core:jackson-core:2.10.2
|    |    |    |    |    \--- com.fasterxml.jackson.core:jackson-databind:2.10.2 (*)
|    |    |    |    +--- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.10.2 (*)
|    |    |    |    \--- com.fasterxml.jackson.module:jackson-module-parameter-names:2.10.2
|    |    |    |         +--- com.fasterxml.jackson.core:jackson-core:2.10.2
|    |    |    |         \--- com.fasterxml.jackson.core:jackson-databind:2.10.2 (*)
|    |    |    +--- org.springframework.boot:spring-boot-starter-tomcat:2.2.5.RELEASE
|    |    |    |    +--- jakarta.annotation:jakarta.annotation-api:1.3.5
|    |    |    |    +--- org.apache.tomcat.embed:tomcat-embed-core:9.0.31
|    |    |    |    |    \--- org.apache.tomcat:tomcat-annotations-api:9.0.31
|    |    |    |    +--- org.apache.tomcat.embed:tomcat-embed-el:9.0.31
|    |    |    |    \--- org.apache.tomcat.embed:tomcat-embed-websocket:9.0.31
|    |    |    |         \--- org.apache.tomcat.embed:tomcat-embed-core:9.0.31 (*)
|    |    |    +--- org.springframework.boot:spring-boot-starter-validation:2.2.5.RELEASE
|    |    |    |    +--- org.springframework.boot:spring-boot-starter:2.2.5.RELEASE (*)
|    |    |    |    +--- jakarta.validation:jakarta.validation-api:2.0.2
|    |    |    |    +--- org.apache.tomcat.embed:tomcat-embed-el:9.0.31
|    |    |    |    \--- org.hibernate.validator:hibernate-validator:6.0.18.Final
|    |    |    |         +--- org.jboss.logging:jboss-logging:3.3.2.Final
|    |    |    |         \--- com.fasterxml:classmate:1.3.4
|    |    |    +--- org.springframework:spring-web:5.2.4.RELEASE (*)
|    |    |    \--- org.springframework:spring-webmvc:5.2.4.RELEASE
|    |    |         +--- org.springframework:spring-aop:5.2.4.RELEASE (*)
|    |    |         +--- org.springframework:spring-beans:5.2.4.RELEASE (*)
|    |    |         +--- org.springframework:spring-context:5.2.4.RELEASE (*)
|    |    |         +--- org.springframework:spring-core:5.2.4.RELEASE (*)
|    |    |         +--- org.springframework:spring-expression:5.2.4.RELEASE (*)
|    |    |         \--- org.springframework:spring-web:5.2.4.RELEASE (*)
|    |    +--- org.springframework.boot:spring-boot-starter-validation:2.2.5.RELEASE (*)
|    |    +--- org.springframework.security:spring-security-crypto:5.2.2.RELEASE
|    |    +--- org.springframework.security:spring-security-rsa:1.0.9.RELEASE
|    |    |    \--- org.bouncycastle:bcpkix-jdk15on:1.64
|    |    |         \--- org.bouncycastle:bcprov-jdk15on:1.64
```

**Comment by wleepang** [comment]
Ah - All of that is from the Spring framework, which is only needed for the `AwsCodeCommitCredentialProvider` class which generates git-credentials via the SigV4 signing method using AWS credentials.  This has nothing to do with using the GRC formatted URLs.  This mechanism is needed for http access to CodeCommit repos.

The source for the class is here:
https://github.com/spring-cloud/spring-cloud-config/blob/master/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/support/AwsCodeCommitCredentialProvider.java

and nearly all of the specific dependencies it has look like they are already part of the classpath.

**Comment by wleepang** [comment]
Latest commits remove the dependency on spring-framework. Also added the ability to use direct "https://" clone urls in addition to GRC formatted clone urls.

File: modules/nextflow/src/main/groovy/nextflow/scm/AwsCodeCommitCredentialProvider.groovy [hunk 5]
```diff
@@ -0,0 +1,320 @@
+package nextflow.scm
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/scm/AwsCodeCommitCredentialProvider.groovy:1]: This source has been copied from the Spring framework lib, isn't it. It's very important to include the original license boilerplate text and author citation.

**Comment by pditommaso** [comment]
Ok, merging in the `feature/codecommit` for now.

**Pull request merged**